minio/minio
# Add MySQL notifications support

Adds a new database target (MySQL, tested with Ver 14.14 Distrib 5.7.17) for notifications.

## Description

The implementation is similar to postgres notifications:

- Notifications are written to a user-specified table (but a specific table structure is required).
- The table is created if it does not exist.
- Event data is stored in a JSON column type.
- Creation of objects results in new rows being inserted if the object name is new, or existing row is updated if an existing object is over-written.
- Deleted objects, cause the corresponding row in the table to be deleted.

In summary, the table will represent all objects in a bucket. 

Additionally the minio server configuration has been bumped up so mysql configuration can be added. Configuration migration code is added with tests.

## Motivation and Context



Adds a feature requested by https://github.com/minio/minio/issues/3818.

## How Has This Been Tested?




Configuration migration related changes have associated test cases updated.

The new target was tested manually. A overview of the process is below:

1. Download mysql docker container
2. Start server with `docker run --name some-mysql -p 3306:3306 -e MYSQL_DATABASE=miniodb -e MYSQL_ROOT_PASSWORD=password -d mysql:latest`
3. Start minio server. It will update configuration version. Then update minio server configuration's mysql section like below and restart the server.
```json
		"mysql": {
			"1": {
				"enable": true,
				"dsnString": "",
				"table": "minio_events",
				"host": "172.17.0.1",
				"port": "3306",
				"user": "root",
				"password": "password",
				"database": "miniodb"
			}
		}

```
4. The server's initial log output will display the ARN with a line like: `SQS ARNs:  arn:minio:sqs:us-east-1:1:mysql`
5. Next use mc to configure events from a bucket to be sent to mysql with something like `mc events add myminio/test arn:minio:sqs:us-east-1:1:mysql`
6. Use mc to copy a file to the server, like `mc cp /etc/issue myminio/test`
7. Connect to mysql server with mysql terminal client with a command similar to `mysql -h 172.17.0.1 -P 3306 -u root -p miniodb`.
8. Type `select * from minio_events` to list the table and you should see the event for the the upload done in step 6, for e.g.:
```
mysql> select * from minio_events;
+------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| key_name   | value                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   |
+------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| test/issue | {"Records": [{"s3": {"bucket": {"arn": "arn:aws:s3:::test", "name": "test", "ownerIdentity": {"principalId": "minio"}}, "object": {"key": "issue", "eTag": "467886be95c8ecfd71a2900e3f461b4f", "size": 26, "sequencer": "14AC142C6D62ABE6"}, "configurationId": "Config", "s3SchemaVersion": "1.0"}, "awsRegion": "us-east-1", "eventName": "s3:ObjectCreated:Put", "eventTime": "2017-03-15T14:22:37Z", "eventSource": "aws:s3", "eventVersion": "2.0", "userIdentity": {"principalId": "minio"}, "responseElements": {"x-amz-request-id": "14AC142C6D62ABE6", "x-minio-origin-endpoint": "http://192.168.0.103:9000"}, "requestParameters": {"sourceIPAddress": "127.0.0.1:58312"}}]} |
+------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.01 sec)
```


## Types of changes


## Checklist:
